presence: Rewrite the backend data model.

mateuszmandera opened the pull request:
This is a continuation of #15030 - I accidentally closed that PR and GitHub doesn't allow me to reopen.

mateuszmandera commented:
Tested reversability of the process on dev environment, by starting on the old model -> moving to the new model -> going back to the old model by reverting the migrations and `git checkout HEAD^^` -> verifying the original UserPresence data seems to be restored fine -> moving to the new model to make sure some unexpected non-idempotent sideeffects leftover from the first time won't mess it up.

mateuszmandera commented on zerver/lib/import_realm.py:
```diff
@@ -352,6 +352,9 @@ def idseq(model_class: Any) -> str:
         return 'zerver_botuserstatedata_id_seq'
     elif model_class == BotConfigData:
         return 'zerver_botuserconfigdata_id_seq'
+    elif model_class == UserPresence:
+        return 'zerver_userpresencenew_id_seq'
```
As a sidenote I wonder if we couldn't and shouldn't detect these automatically with some appropriate postgres query, rather than hard-coding special cases in this function. This is a very easy to detail miss (unless tests catch the error of course, which they generally will, but still...)

mateuszmandera commented:
Reminder note to update `performance.md` Presence section once this this is finished and merged

timabbott commented:
I did a rebase of the migrations of this in preparation for trying to deploy this on chat.zulip.org.  I also did a careful read through the migration, and added a proposed commit that substantially simplifies the database migration, and I think is the best we can do.  (When thinking about the cases where we set `last_active_time` and `last_connected_time` separately, I realized that the values computed by that approach were likely to be erroneous with similar magnitude as the errors this simpler algorithm generates, and this simpler algorithm will be easier to reason about).  @mateuszmandera can you review that proposed decision?

One thing that needs verification is whether the `order_by` + `distinct` does the right thing (i.e. does it select via the `DISTINCT` always the latest row).  @mateuszmandera can you figure out a way to confirm that?

I also tested that the original version of this migration ran in 6s with 23k `UserPresence` records chat.zulip.org; I think that is fast enough that we can likely get away with not doing a far fancier series of migrations to minimize downtime, as long as we're confident we won't want to revert the conversion.  That said, I think it could be worth doing the migration sequence as (1) rename `UserPresence` to `UserPresenceOld`, (2) create new `UserPresence`, (3) backfill UserPresence => UserPresenceOld, (4) drop UserPresenceOld, with the thinking being that doing so would allow us to start the updated server (with the `UserPresence` queue worker disabled) while the backfill migration (3) is running, and thus drop fewer HTTP requests.  Effectively, that'd mean that we'd replay any presence events that come in while the backfill migration is running just after the backfill finishes.  Clients would end up with accurate presence information again within a minute after the queue worker catches up, I think.

@alexmv as well for thoughts on that migration strategy.

The other thing that this PR will need is a careful review of all the time offset constants for whether they make sense logically, and document the decisions in comments/ReadTheDocs so that we understand how presence works.  I expect to do that process before deploying it to chat.zulip.org hopefully later this week.

mateuszmandera commented:
> One thing that needs verification is whether the `order_by` + `distinct` does the right thing (i.e. does it select via the `DISTINCT` always the latest row). @mateuszmandera can you figure out a way to confirm that?

Yup, that's detailed in https://docs.djangoproject.com/en/3.2/ref/models/querysets/#distinct:

```
When you specify field names, you must provide an order_by() in the QuerySet, and the fields in order_by() must start with the fields in distinct(), in the same order.

For example, SELECT DISTINCT ON (a) gives you the first row for each value in column a. If you don’t specify an order, you’ll get some arbitrary row.
```

mateuszmandera commented on zerver/migrations/0353_userpresencenew_fill_data.py:
```diff
@@ -63,6 +63,8 @@ class Migration(migrations.Migration):
     Ports data from the UserPresence model into the new one.
     """
 
+    atomic = False
```
I think we want this, to really have batching in that `bulk_update` call, right?

timabbott replied:
Yeah, that seems correct.  I guess the downside of doing that is that new UserPresence objects might be created in the meantime; we probably can just ignore that race since a normal client submits presence data every minute (and disabling the queue worker as discussed above would prevent such a race).

mateuszmandera replied:
Hmm, I don't think `atomic` a
influences this issue at all?

timabbott replied:
No -- sorry I'd been thinking about switching this to a different loop structure that we write ourselves that could catch objects created while the migration is running.

timabbott commented:
@mateuszmandera do you want to attempt the reordering/restructuring of the migrations that I proposed, with the `presence_old` approach?

mateuszmandera commented:
> @mateuszmandera do you want to attempt the reordering/restructuring of the migrations that I proposed, with the `presence_old` approach?

Sure :smile:

mateuszmandera commented on zerver/models.py:
```diff
@@ -3399,78 +3399,43 @@ class UserPresence(models.Model):
         ]
 
     id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name="ID")
     user_profile: UserProfile = models.ForeignKey(UserProfile, on_delete=CASCADE)
```
Hmm, do we perhaps want this to be either `unique=True` or `OneToOneField`?

timabbott replied:
Yes, I think `unique=True` is missing here.

mateuszmandera commented:
@timabbott Done - can you review?
Notes:
- The first migration got kind of fiddly due to needing to manually rename indexes etc. to avoid name conflicts (elaborated in a comment in the migration).
- I've done a very basic test of the deployment scheme in dev env:
  1. Start migrations with the backfill migrations delibarely slow with `time.sleep(180)` in the middle of it.
  2. `run-dev.py` with the user_presence worker disabled by commenting out its `@assign_queue`.
  3. Do some activity in the webapp and stop the server.
  4. Migrations finishes. Verify new presence events are sitting in the queue.
  5. Start run-dev again, with the queue on, confirm it processes the events to catch up.
  6. Confirm logging in and out as various users is reflected in their Active status in the webapp.
 - added a paragraph about this method of deployment in the first commit message
 - https://github.com/zulip/zulip/pull/16381#discussion_r714981472 remains a questions - are we missing a uniqueness constraint?
 - added `[squash] import_realm: Fix special index name handling block.` since that old block is wrong with this new migration approach. It seems that Django may have changed the naming pattern for new tables? If so, `zerver.lib.import_realm.idseq` will need to be adjusted for this somehow soon
 - Opened #19835 for a weird `renumber-migrations` bug that sneakily altered my SQL when rebasing :neutral_face: I've rebased on top of latest `upstream/main` but if you end up rebasing on top of some new migrations and running `renumber-migrations` when deploying, watch out for this bug since it'll make the migration fail

mateuszmandera commented:
@timabbott Done - can you review?
Notes:
- The first migration got kind of fiddly due to needing to manually rename indexes etc. to avoid name conflicts (elaborated in a comment in the migration).
- I've done a very basic test of the deployment scheme in dev env:
  1. Start migrations with the backfill migrations delibarely slow with `time.sleep(180)` in the middle of it.
  2. `run-dev.py` with the user_presence worker disabled by commenting out its `@assign_queue`.
  3. Do some activity in the webapp and stop the server.
  4. Migrations finishes. Verify new presence events are sitting in the queue.
  5. Start run-dev again, with the queue on, confirm it processes the events to catch up.
  6. Confirm logging in and out as various users is reflected in their Active status in the webapp.
 - added a paragraph about this method of deployment in the first commit message
 - https://github.com/zulip/zulip/pull/16381#discussion_r714981472 remains a questions - are we missing a uniqueness constraint?
 - added `[squash] import_realm: Adjust UserPresence id sequence name.` since the name has changed with this new migration approach. 
 - Opened #19835 for a weird `renumber-migrations` bug that sneakily altered my SQL when rebasing :neutral_face: I've rebased on top of latest `upstream/main` but if you end up rebasing on top of some new migrations and running `renumber-migrations` when deploying, watch out for this bug since it'll make the migration fail

timabbott commented on zerver/lib/actions.py:
```diff
@@ -6885,10 +6898,8 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
-            timestamp__gte=recent,
+            last_active_time__gte=recent,
         )
-        .exclude(client__name="ZulipMobile")
```
This needs to remove the `ZulipMobile` discussion in the comment above.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5451,7 +5452,8 @@ def send_presence_changed(user_profile: UserProfile, presence: UserPresence) ->
         return
 
-    presence_dict = presence.to_dict()
+    # The mobile app handles these events so we need to use the old format.
```
We probably want to extend this comment to mention that we should make the `slim_presence` client capability control the format of these events once that is fully implemented.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5481,39 +5483,50 @@ def do_update_user_presence(
+        last_active_time=log_time,
+        last_connected_time=log_time,
         realm_id=user_profile.realm_id,
     )
```
We probably want a comment here since this can look like a bug: we're only setting `last_active_time` unconditional of the reported status because we don't have an existing `UserPresence` row at all.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5481,39 +5483,50 @@ def do_update_user_presence(
+    # have newer presence than the reported log_time.
+    if (
+        not created
+        and log_time - presence.last_active_time > datetime.timedelta(seconds=55)
```
Presumably this 55 should be some sort of constant defined in settings that's effectively our poll frequency.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5481,39 +5483,50 @@ def do_update_user_presence(
+    if (
+        not created
+        and status == UserPresence.LEGACY_STATUS_ACTIVE_INT
+        and log_time > presence.last_active_time
```
Why does this one not have the `delta > 55 seconds` check that we have for `last_connected_time`?

mateuszmandera replied:
Yeah I think that was a bug, we don't want to be writing to db on every ACTIVE ping, though I originally thought that may be the intent. Also fixed https://github.com/zulip/zulip/pull/16381/commits/a978d12c0e58a5b3409b69d02e668bce27709ade#r716084413 in the `if` above.

timabbott commented on zerver/lib/presence.py:
```diff
@@ -112,22 +88,35 @@ def get_legacy_user_info(
+    and is not meant to be used on old presence data.
+    """
+    if (
+        presence.last_active_time + datetime.timedelta(minutes=1, seconds=10)
```
This constant should probably be extracted to a value in `computed_settings.py` as well.

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -5512,17 +5518,16 @@ def do_update_user_presence(
     # have newer presence than the reported log_time.
-    if (
-        not created
-        and log_time - presence.last_active_time > datetime.timedelta(seconds=55)
```
I think it was a bug to be looking at timedelta from `last_active_time` here. It would have meant updating the UserPresence row each time we get a new IDLE ping, if last_active_time was more than 55s ago.

mateuszmandera commented on zerver/lib/presence.py:
```diff
@@ -35,72 +24,60 @@ def get_status_dicts_for_rows(
+    #
+    # We expect this code to remain mostly unchanged until we can delete it.
+
+    if timezone_now() - last_active_time > datetime.timedelta(minutes=2):
```
I think this `timedelta` may be wrong? Shouldn't this use the same value as
```
def format_legacy_presence_dict(presence: UserPresence) -> Dict[str, Any]:
    """
    This function assumes it's being called right after the presence object was updated,
    and is not meant to be used on old presence data.
    """
    if (
        presence.last_active_time + datetime.timedelta(minutes=1, seconds=10)
        presence.last_active_time
        + datetime.timedelta(seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS)
        >= presence.last_connected_time
    ):
        status = UserPresence.LEGACY_STATUS_ACTIVE
        timestamp = datetime_to_timestamp(presence.last_active_time)
    else:
        status = UserPresence.LEGACY_STATUS_IDLE
        timestamp = datetime_to_timestamp(presence.last_connected_time)
```

which we've set to 70seconds?

timabbott replied:
So I think the core problem statement here is the concern that you have multiple clients, and one is reporting IDLE but another is reporting ACTIVE, and we should not be overriding the "I'm active" information from the computer you're at with information from another client that thinks you're IDLE. (The new `last_connected_time`/`last_active_time` system is **much** cleaner on this point).

So I'm pretty sure the meaning of that 70s was basically "If you have a presence update showing you were IDLE within the last PRESENCE_UPDATE_MIN_FREQ_SECONDS + slop, but none showing you are ACTIVE, then we know there isn't another client that is reliably updating us with the information that you're active. 

I think if we assume that network failures don't happen, then indeed it doesn't matter whether we're doing this check just after the presence object was updated -- the threshold should always be.
`settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS`. And I think it's pretty low harm for you to switch to being shown as IDLE (with a more current connection) if your computer fails to tell the server that you're online, so I don't think we need to add an extra round trip of slop here. 

Does that mean we can just call `format_legacy_presence_dict` here? I guess it's expecting a `UserPresence` object rather than two values, but that can probably be extracted.

mateuszmandera replied:
https://github.com/zulip/zulip/pull/16381/commits/464c33b1a671f9b5747c0c3618c4b69f2f332a41#r1157752227

mateuszmandera commented:
@timabbott Ready for review. Posted a couple of comments pushed as a new `[squash]` to make it clear what's changing.

mateuszmandera commented:
@timabbott Rebased this. Most notable change was that I ended up adding this block in `do_update_user_presence`  to fix a bug detected in `test_away_events`:
```
    if status == UserPresence.LEGACY_STATUS_IDLE_INT:
        # If the presence entry for the user is just to be created, and
        # we want it to be created as idle, then there needs to be an appropriate
        # offset between last_active_time and last_connected_time, since that's
        # what the event-sending system calculates the status based on, via
        # format_legacy_presence_dict.
        defaults["last_active_time"] = log_time - datetime.timedelta(
            seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS + 1
        )
```

The bug was that in the `presence_enabled` codepath in `do_change_user_setting` when disabling presence, the `UserPresence` for the user is deleted to be replaced by a new `UserPresence` with slightly-stale timestamps. But as `test_away_events` showed, this lead to an event being sent  showing `status: "active"` still for the user, when it's expected to be `idle`. That's because `format_legacy_presence_dict` does calculations based on the presence-object only, without looking at current time (which is intended i believe). So the best way for "idleness" to be reflected in the `UserPresence` being created is by having an offset between the timestamps.

timabbott commented on zerver/actions/presence.py:
```diff
@@ -117,17 +123,16 @@ def do_update_user_presence(
+        > datetime.timedelta(seconds=settings.PRESENCE_UPDATE_MIN_FREQ_SECONDS)
     ):
         presence.last_active_time = log_time
         update_fields.append("last_active_time")
```
Do we want to update `last_connected_time` whenever we're also updating `last_active_time`? Like, have a `if log_time > presence.last_connected_time: ` block here to update `last_connected_time` too?

I feel like that would make it easier to reason about things -- we could promise to downstream users of the system that `last_connected_time >= last_active_time`, which I think this implementation as written does not achieve.

mateuszmandera replied:
Good point, let's keep that property - updated. Would we want to have a db constraint for this to detect violations? I guess not worth it, but just throwing the idea - something like `alter table zerver_userpresence add constraint  X check (last_connected_time >= last_active_time);`

timabbott replied:
I think given that we only write to the table in this one function, that is probably not necessary, but if it's easy to add via the Django constraint system, it might be a good way to catch bugs.

timabbott commented on zproject/default_settings.py:
```diff
@@ -494,6 +494,18 @@
 USER_LIMIT_FOR_SENDING_PRESENCE_UPDATE_EVENTS = 100
 
+# Controls the how much newer a user presence update needs to be
+# than the currently saved last_active_time in order for us to
```
I think the reference to `last_active_time` is no longer accurate? Maybe "the currently saved last_active_time or last_connected_time..." would work?

timabbott commented on zerver/actions/presence.py:
```diff
@@ -77,39 +81,62 @@ def do_update_user_presence(
+    time_since_last_active = log_time - presence.last_active_time
+    now_online = time_since_last_active > datetime.timedelta(
+        seconds=settings.OFFLINE_THRESHOLD_SECS - 10
+    )
```
Hmm, is this `-10` logic correct? I think the idea is we want to immediately notify clients if they would have previously shown the user as offline. It's weird -- if we did exactly OFFLINE_THRESHOLD_SECONDS, that'd be wrong in that we might fail to notify clients that were about to start displaying the user as offline. At the same time, we only want this code path to send a presence update via the events system if the user's state has actually changed.

Do we want `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL`, in other words, "if a client might not otherwise display you as offline before the next time they ping the server"? I worry that might be so small that this code path triggers incorrectly really often.

mateuszmandera replied:
> Do we want `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL`, in other words, "if a client might not otherwise display you as offline before the next time they ping the server"? I worry that might be so small that this code path triggers incorrectly really often.

`OFFLINE_THRESHOLD_SECONDS` is 5 minutes though, and `ACTIVE_PING_INTERVAL_MS = 50 * 1000` (50 seconds) so that delta is actually 4 minutes 10 seconds so not short at all. So perhaps that would be fine, unless I'm thinking about this wrong and that is too large instead?

timabbott replied:
OK if `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL - 10` is 4 minutes, that's probably very reasonable for this -- it should mean that a client only triggers this if it's missed like 4 pings in a row, which is a good definitely for "was previously offline".

Oh, but the frontend has `OFFLINE_THRESHOLD_SECS=140`; and we had that hardcoded in `message_send.py` too. I think this is all good IF we sort out that discrepancy. I like the idea of this being a setting long-term, since it's a thing we can maybe tune in Zulip Cloud during an incident or something, so we may want to put it into the `/server_settings` API at some point. 

```
$ git grep OFFLINE_THRESHOLD_SECS
docs/subsystems/presence.md:  useful sample code; the `OFFLINE_THRESHOLD_SECS` check is critical
frontend_tests/node_tests/presence.js:const OFFLINE_THRESHOLD_SECS = 140;
frontend_tests/node_tests/presence.js:        active_timestamp: now - OFFLINE_THRESHOLD_SECS / 2,
frontend_tests/node_tests/presence.js:        active_timestamp: now - OFFLINE_THRESHOLD_SECS * 2,
frontend_tests/node_tests/presence.js:        idle_timestamp: now - OFFLINE_THRESHOLD_SECS / 2,
frontend_tests/node_tests/presence.js:    const recent = now + 1 - OFFLINE_THRESHOLD_SECS;
frontend_tests/node_tests/presence.js:    const a_while_ago = now - OFFLINE_THRESHOLD_SECS * 2;
static/js/presence.js:const OFFLINE_THRESHOLD_SECS = 140;
static/js/presence.js:    if (age(active_timestamp) < OFFLINE_THRESHOLD_SECS) {
static/js/presence.js:    if (age(idle_timestamp) < OFFLINE_THRESHOLD_SECS) {
zerver/actions/message_send.py:    OFFLINE_THRESHOLD_SECS = 140
zerver/actions/message_send.py:    recent = timezone_now() - datetime.timedelta(seconds=OFFLINE_THRESHOLD_SECS)
zerver/actions/user_settings.py:                seconds=settings.OFFLINE_THRESHOLD_SECS + 120
zerver/tests/test_presence.py:        with self.settings(OFFLINE_THRESHOLD_SECS=1):
zerver/views/presence.py:    if aggr_status_duration > settings.OFFLINE_THRESHOLD_SECS:
zproject/default_settings.py:OFFLINE_THRESHOLD_SECS = 5 * 60
```

It looks like the randomly choosing between those two in values in different places; but I think it may be the 140s is the operative one today, in that `status_from_raw` mostly looks at the timestamps, not the server `offline` flag...

timabbott replied:
OK, I took the time to analyze the history here, and posted a plan in the thread here:

The proposal in https://chat.zulip.org/#narrow/stream/3-backend/topic/UserPresence.20system.20rewrite/near/1506240

With that plan, this line can be `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL - 10` which would be 130s -- long enough that we'd need to have missed two pings in a row; that seems safe and correct to me. We may want to do something to monitor that this doesn't happen excessively often; but that should be visible in the event logs.

mateuszmandera commented:
@timabbott I squashed commits, addressed the two comments above, responded to https://github.com/zulip/zulip/pull/16381#discussion_r1087192633 without changing anything because I don't feel sure i understand this parameter correctly - also made the migration changes (getting rid of the special case name for the id sequence and making the more efficient `latest_presence_per_user` query) @alexmv pointed out in https://chat.zulip.org/#narrow/stream/3-backend/topic/UserPresence.20system.20rewrite

alexmv commented on zerver/migrations/0425_userpresence_fill_data.py:
```diff
@@ -0,0 +1,62 @@
+    # unlikely to be significant in practice, and in any case is an
+    # unavoidable flaw caused by the legacy previous data model.
+    with connection.cursor() as cursor:
+        cursor.execute(
+            "SELECT realm_id, user_profile_id, MAX(timestamp) FROM zerver_userpresenceold WHERE status IN (1, 2) GROUP BY realm_id, user_profile_id"
+        )
+        latest_presence_per_user = cursor.fetchall()
```
This is doable directly in Django if you prefer:
```python3
UserPresenceOld.objects.values('realm_id','user_profile_id').annotate(latest=Max('timestamp')).filter(status__in=[1,2])
```

mateuszmandera replied:
I can swap it to the ORM query, though I went with SQL on purpose because I somehow don't find it very understandable into what `GROUP BY` SQL exactly Django will translate `.annotate` queries - and with raw SQL we can know exactly what it's going to do :neutral_face:

timabbott replied:
Yeah it's nice to use the ORM, but I also find Django's annotate syntax a little hard to read, so don't have a strong opinion on this one way or another.

timabbott commented on zerver/models.py:
```diff
@@ -4020,77 +4020,42 @@ class UserPresence(models.Model):
+    # interacting with a computer running the desktop app)
+    last_active_time = models.DateTimeField(default=timezone_now, db_index=True)
+
+    # The following constants are used in the presence API for
+    # communicating whether a user is active (last_active_time recent)
+    # or idle (last_connected_time recent) or offline (neither
+    # recent).  They're no longer part of the data model.
+    LEGACY_STATUS_ACTIVE = "active"
+    LEGACY_STATUS_IDLE = "idle"
+    LEGACY_STATUS_ACTIVE_INT = 1
+    LEGACY_STATUS_IDLE_INT = 2
```
Our intent is for these to only remain in the API for backwards-compatibility with older clients that don't know about the upcoming cleaner presence API. Soon after this PR, we'll be changing the API to just send `last_active_time` and `last_connected_time` values for each user to modern clients; so I think the name is appropriate.

mateuszmandera commented on zerver/actions/message_send.py:
```diff
@@ -714,10 +708,8 @@ def filter_presence_idle_user_ids(user_ids: Set[int]) -> List[int]:
+            last_active_time__gte=recent,
         )
-        .exclude(client__name="ZulipMobile")
         .distinct("user_profile_id")
```
True, tweaked!

mateuszmandera commented:
Rebased a tweaked.

alexmv commented on zilencer/management/commands/populate_db.py:
```diff
@@ -795,17 +795,12 @@ def assign_time_zone_by_delivery_email(delivery_email: str, new_time_zone: str)
                 UserPresence.objects.get_or_create(
                     user_profile=user,
                     realm_id=user.realm_id,
-                    client=client,
-                    timestamp=date,
-                    status=status,
+                    last_active_time=date,
+                    last_connected_time=date,
                 )
```
```suggestion
                UserPresence.objects.get_or_create(
                    user_profile=user,
                    realm_id=user.realm_id,
                    defaults={"last_active_time": date, "last_connected_time": date},
                )
```

mateuszmandera replied:
Thanks, tweaked!

mateuszmandera commented:
Uh, commits got messed up in the rebasing process, do not merge until fixed

mateuszmandera commented:
Fixed

mateuszmandera commented on zerver/actions/presence.py:
```diff
@@ -77,39 +81,73 @@ def do_update_user_presence(
-    became_online = (status == UserPresence.ACTIVE) and (stale_status or was_idle)
+    time_since_last_active = log_time - presence.last_active_time
+    now_online = time_since_last_active > datetime.timedelta(
+        # With the default tuning of these constants, where OFFLINE_THRESHOLD_SECS
+        # is equal to PRESENCE_PING_INTERVAL_SECS * 3 + 20, the expression below
+        # is equivalent to just 2 * PRESENCE_PING_INTERVAL_SECS + 10, so just enough
+        # time for the client to have missed two pings plus a small extra margin.
+        seconds=settings.OFFLINE_THRESHOLD_SECS
+        - settings.PRESENCE_PING_INTERVAL_SECS
+        - 10
+    )
```
Tweaked this to address https://github.com/zulip/zulip/pull/16381/commits/135d46f2c78c1679bb46ccbab0f58df295073ca1#r1101923300 and attempted to add a comment giving some clarification, so that this doesn't become a mysterious magical arithmetic expression in the future - but not sure what I wrote there is enough - feel free to perhaps add more justification if appropriate

alexmv replied:
Not having followed this conversation, this comment leaves me wondering "so why not write `2 * PRESENCE_PING_INTERVAL_SECS + 10`?"

timabbott replied:
I think the reason we're using this formula is that we're sending this notice if the client is at risk of demoting the user to offline before the next time it would query the server via a regular checkin. Probably the comment should reflect that.

timabbott replied:
Maybe:

> Here, we decide whether the user is newly online, and we should consider sending an immediate presence update via the events system that this user is now online, rather than waiting for other clients to poll the presence update. Immediate presence updates add load to the system, so we only want to do this if the user has missed a couple regular presence checkins (so their state is at least `2 * PRESENCE_PING_INTERVAL_SECS + 10` old), and also is risk of being shown by clients as offline before their next regular presence checkin (so at least `settings.OFFLINE_THRESHOLD_SECS - settings.PRESENCE_PING_INTERVAL_SECS - 10`). These two values happen to be the same in the default configuration.

Does it make sense to write this block using `max(2 * PRESENCE_PING_INTERVAL_SECS + 10, settings.OFFLINE_THRESHOLD_SECS - settings.PRESENCE_PING_INTERVAL_SECS - 10)`? 

Or maybe better we should write `assert OFFLINE_THRESHOLD_SECS >= 3 * settings.PRESENCE_PING_INTERVAL_SECS + 20` here next to a comment of this form and leave it roughly as written; I think it would not be a happy state if that were not true given this behavior.

mateuszmandera replied:
Added the comment with tiny phrasing tweaks, and went with adding the assert. Kinda ugly with the exclusion for tests:
```
    if not settings.TEST_SUITE:  # nocoverage
        # We exclude the test suite from the assertions, since in tests we may want to
        # mock settings arbitrarily.
        assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
```

the exclusion is not really necessary since it's only one existing test that breaks the assumption so it could be fixed - but I feel this might be the right choice in general so that any future tests can be written mocking these values freely without crashing into an annoying assertion?

timabbott replied:
I don't like `not TEST_SUITE` assertions since they make the test suite not predictive of what will happen in production. In particular, if we letter changes these values and don't remember this line of code, we could merge code that fails an assertion every time it runs :)

It would be better to change the test, and have a comment here the explains what we're doing with the math.

mateuszmandera replied:
Cool, the test didn't really need to  any settings, so just squashed this tweak into the original commit:
```diff
diff --git a/zerver/actions/presence.py b/zerver/actions/presence.py
index cf480c6076..392639b9f7 100644
--- a/zerver/actions/presence.py
+++ b/zerver/actions/presence.py
@@ -120,10 +120,7 @@ def do_update_user_presence(
         # Same approach as above.
         time_since_last_connected = datetime.timedelta(days=1)
 
-    if not settings.TEST_SUITE:  # nocoverage
-        # We exclude the test suite from the assertions, since in tests we may want to
-        # mock settings arbitrarily.
-        assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
+    assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
     now_online = time_since_last_active > datetime.timedelta(
         # Here, we decide whether the user is newly online, and we need to consider
         # sending an immediate presence update via the events system that this user is now online,
diff --git a/zerver/tests/test_presence.py b/zerver/tests/test_presence.py
index b2f0a0c72e..d3767d4a83 100644
--- a/zerver/tests/test_presence.py
+++ b/zerver/tests/test_presence.py
@@ -630,8 +630,17 @@ class UserPresenceAggregationTests(ZulipTestCase):
         user = self.example_user("othello")
         self.login_user(user)
         validate_time = timezone_now()
-        with self.settings(OFFLINE_THRESHOLD_SECS=1):
-            result_dict = self._send_presence_for_aggregated_tests(user, "idle", validate_time)
+        result_dict = self._send_presence_for_aggregated_tests(user, "idle", validate_time)
+
+        with mock.patch(
+            "zerver.views.presence.timezone_now",
+            return_value=validate_time + datetime.timedelta(settings.OFFLINE_THRESHOLD_SECS + 1),
+        ):
+            # After settings.OFFLINE_THRESHOLD_SECS + 1 this generated, recent presence data
+            # will count as offline.
+            result = self.client_get(f"/json/users/{user.email}/presence")
+        result_dict = self.assert_json_success(result)
+
         self.assertDictEqual(
             result_dict["presence"]["aggregated"],
             {

```

timabbott commented on zerver/actions/presence.py:
```diff
@@ -77,39 +81,73 @@ def do_update_user_presence(
+        # format_legacy_presence_dict.
+        defaults["last_active_time"] = log_time - datetime.timedelta(
+            seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS + 1
+        )
```
Hmm, I wonder if it'd be better to address the TODO above by making `None` be a supported value for both `last_connected_time` and `last_active_time`? That would affect the future API, but it might be appropriate for both the case in this block (we'd set `last_active_time=None`, but `last_connected_time` would be set as `log_time`) and also the "New account created" block.

Would require some adjustments.

mateuszmandera replied:
@timabbott Yeah we should add that option though I'm thinking this should probably be done as an iteration on this after we test the current state of this on czo? This will add some more complexity, so that might make it easier to review and identify the source of bugs if we run into any on czo?

mateuszmandera commented on zerver/lib/presence.py:
```diff
@@ -48,36 +48,18 @@ def get_modern_user_presence_info(
-
-    # This field was never used by clients of the legacy API, so we
-    # just set it to a fixed value for API format compatibility.
-    pushable = False
```
Did this deduplication to resolve https://github.com/zulip/zulip/pull/16381#discussion_r716084830

mateuszmandera commented:
Pushed a proposed commit for supporting the `None` values. A couple of notes on additional things I stumbled onto:

1. The "get a single user's presence" API endpoint `/users//<user id or email>/presence` is kind of weird? Perhaps because it's mostly unused? It makes this adjustment to the info dics it returns:
https://github.com/zulip/zulip/blob/946b4e73ca6dbc34788b8799d9d8de04a2b17809/zerver/views/presence.py#L55

Which would make sense on its own, but `/realm/presence` does not make such an adjustment, so that's an inconsistency. 173f34b7aa39d4224388c7010cc382ac762f38fc is the commit that added that stuff, but I just feel I'm missing some history here since I don't really understand the whole `aggregated` presence info concept beyond it being some legacy API leftover. Perhaps there's nothing to be done here right now and we'll clean it up in the future when abandoning these legacy APIs?

2. Just want to also flag that when creating legacy (non-slim) presence info dicts for API endpoints like `/realm/presence`  in `get_presence_dicts_for_rows`, the `get_legacy_user_presence_info` function calculates the `status` value purely based on the timestamps of the `UserPresence` objects, with no regard for whether they are stale, or fresh. That is, it will e.g. report `active` status if `last_active_time` is fresher than `last_connected_time` even if those timestamps are weeks old. However this is consistent, with how the presence system has been doing it so far in `main`, because as you can see in https://github.com/zulip/zulip/blob/946b4e73ca6dbc34788b8799d9d8de04a2b17809/zerver/lib/presence.py#L76 there also the `status` is just set based on the existing `UserPresence` records with no regard for whether they're very recent or old. My understanding is that this is fine, because clients aren't supposed to look at the `status` from these APIs anymore and should be calculating the status on their based on the `timestamp` from the API only?

mateuszmandera commented on zerver/lib/presence.py:
```diff
@@ -26,15 +26,40 @@ def get_presence_dicts_for_rows(
+    if dt is None:
+        return date_joined
+
+    return dt
```
This function may be a bit pointless, but it lets us avoid having lots of annoying `if`s in the code and drives home the idea that `date_joined` is to be used as the fallback if the presence timestamp is `None`  (instead of that decision being made in multiple places separately, allowing inconsistencies in the future if this behavior is changed in one place but not the other) and provides a nice place to just have a docstring explaining what's going on.

timabbott replied:
This lgtm though I'd maybe call it "the legacy API" somewhere, just so this comment makes sense once we've changed the main API to something better.

(Also you should generally line-wrap prose at ~70 characters, rather than the 100 that Black will do; this avoids the GitHub scrollbar above and is easier to read in other settings too).

mateuszmandera replied:
Tweaked the docstring to mention "legacy API" in one place and changed the docstring  to have max line length 70. Aside of that, these latest changes are still waiting for review.

timabbott commented:
> Perhaps there's nothing to be done here right now and we'll clean it up in the future when abandoning these legacy APIs?

Yeah, I think that's right -- basically by the time we'd added the individual user presence endpoint, we'd decided the multiple-clients stuff was a bad idea, and so we were just trying to hide that complexity from users. The `aggregated` thing was just "the most recent of the various' clients' states", aka the thing that we expect all clients to actually look at with the old API.

>  My understanding is that this is fine, because clients aren't supposed to look at the status from these APIs return anymore and should be calculating the status on their based on the timestamp from the API only?

Yes, that's correct -- the previous API gave you, awkwardly, "this user was online at time X" and the client does the math to see if that means they're currently online/offline or not. The plan is for the new API to give you that same information, but not awkwardly :)

mateuszmandera commented:
Oops, addressed that outstanding comment in https://github.com/zulip/zulip/pull/16381#discussion_r1164649072

timabbott commented:
Looks great, we can merge to `main` once we've verified this on chat.zulip.org for a bit.

mateuszmandera commented:
Quick rebase

timabbott commented:
We probably need another rebase to renumber migrations and API changes after merged PRs -- and I think the only other thing we're waiting for is the hack to include `default` alongside `aggregated`?

mateuszmandera commented:
@timabbott Pushed a commit to add a copy of `aggregated` in the presence info response dicts like discussed in #api design. Is this how we wanted to do this?

mateuszmandera commented on zerver/lib/presence.py:
```diff
@@ -99,6 +99,11 @@ def get_legacy_user_presence_info(
+    # format under some client name.
+    result["default"] = result["aggregated"].copy()
+
     result["website"] = most_recent_info
```
Also just wanted to highlight here that we are putting new-format presence info data under the `website` key, which is kind of wonky - in case we want to make any further changes here

mateuszmandera replied:
I think not clear to me why what we put in `result["website"] ` is not sufficient here. Should we perhaps doing just that and calling that `default` instead of `website`?

timabbott replied:
Yeah, it sounds like we had already thought of this detail and just picked a different name. Let me test-deploy this without this final commit, and we can discuss in "api design".

timabbott commented:
I decided to leave this for @alexmv to test-deploy tomorrow, but I think we should do so dropping the last commit -- there's no need to have 3 blobs in each object.

timabbott commented:
Rebased for `statsd` having been removed; I also added an API feature level bump, since it seems important enough of a change to document.

With those changes, marked to merge once CI passes.
